### PR TITLE
Extractor with authentication

### LINKED ISSUE #3157: Extractor: read and use authentication token
Extractor should read the access token from a configurable location and use it to work against an authenticated sandbox.

Extractor should also try to re-read the stored token if a loaded one expires.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
As per title, this PR makes Extractor works against a Ledger API that requires authentication.

The `RetryStrategy` has been moved out of the Ledger API Test Tool and into its own library.

Since we already have `libs-haskell`, I decided to put it in a newly created `libs-scala` where I also moved `//ledger/ledger-api-grpc-utils` (and renamed them `grpc-utils`). Please pay extra attention to this in the review to ensure I didn't get anything wrong regarding publishing on Maven.

The tests regarding expiring tokens on running streams have long wait time (five seconds between each operation) to lower the probability of CI flakes. It adds ~50s to the build time but they should run only when the Bazel caches misses an un-to-date Extractor.

Closes #3157.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    ledgerId = None,
+    participantId = None,
+    applicationId = None,
+    exp = None,
+    admin = true,
+    actAs = List(operator),
+    readAs = List(operator)
+  )
+
+  private val token = Some(Files.createTempFile("Extractor", "AuthSpec"))
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala]: Mostly to match the signature of the methods taking a token, but I can push it to the call site.

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        Failure(e)
+    }
+
+}
+
+final class TokenHolder(location: Option[Path]) {
+
+  import TokenHolder.slurpToken
+
+  private[this] val ref = location.map(path => new AtomicReference(slurpToken(path)))
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: The idea was to make `TokenHolder` handle the fact that the token is optional transparent to the caller but I see your point, I'll refactor this.

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import org.slf4j.LoggerFactory
+
+import scala.util.{Failure, Try}
+import scala.util.control.NonFatal
+
+object TokenHolder {
+
+  private val logger = LoggerFactory.getLogger(classOf[TokenHolder])
+
+  private def slurpToken(path: Path): Try[String] =
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Yeah, I wrestled with the internal a bit and ended up with this, your point makes sense, working on it, thanks.

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    Try {
+      logger.info(s"Reading token from $path...")
+      Files.readAllLines(path).stream.collect(Collectors.joining("\n"))
+    } recoverWith {
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: I'd rather do the following to also go in a direction that covers your other comments.

```scala
  private def slurpToken(path: Path): Option[String] =
    try {
      logger.info(s"Reading token from $path...")
      Option(Files.readAllLines(path).stream.collect(Collectors.joining("\n")))
    } catch {
      case NonFatal(e) =>
        logger.error(s"Unable to read token from $path", e)
        None
    }
```

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Ok, I moved the path optionality out of here, the caller will deal with that.

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (gerolf-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: 👍

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerolf-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: 👍

File: extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    val process = extractor(tailWithAuth)
+    setToken(operatorPayload.expiresInFiveSeconds.asHeader())
+    Delayed.Future.by(10.seconds) {
+      newSyncClient.submitAndWait(SubmitAndWaitRequest(commands = dummyRequest.commands))
+    }
+    Delayed.by(15.seconds)(setToken(operatorPayload.asHeader()))
+    Delayed.Future.by(20.seconds) {
+      newSyncClient.submitAndWait(SubmitAndWaitRequest(commands = dummyRequest.commands))
+    }
+    Delayed.Future.by(25.seconds)(process.shutdown().map(_ => succeed))
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala]: Yes. In exploring this point I realized that the tests were *completely* broken. I was not issuing commands with the access tokens set, I didn't event call `process.run()` and was relying on the outcome of `process.shutdown()` (which simply returns a success when the process is shut down - that is, always). I've re-written the last two tests as a single one, using an ad-hoc writer to inspect the state of what's written there. It's a bit complex and I'd ask you to double check it for correctness. :pray:

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/b73de1b433351ccb87b150f47f9574a5d87860af

[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/461cf6189ae0521d5fb616d54dddc3a3ed3b97f5

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/461cf6189ae0521d5fb616d54dddc3a3ed3b97f5

File: extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/helpers/TokenHolder.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/461cf6189ae0521d5fb616d54dddc3a3ed3b97f5

File: extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/27fe3b2aa9a340c4a5d07cfab375ce877b8c2c73

File: extractor/src/main/scala/com/digitalasset/extractor/config/ConfigParser.scala [hunk 6]
```diff
@@ -244,6 +244,13 @@ object ConfigParser {
           c.copy(tlsCaCrt = Some(path))
         }
 
+      note("\nAuthentication:")
+
+      opt[String]("access-token-file")
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/main/scala/com/digitalasset/extractor/config/ConfigParser.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/e7905b150880f9a3e20c811ffed5d753338544b5

File: extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply extractor/src/test/suite/scala/com/digitalasset/extractor/AuthSpec.scala]: Addressed by https://github.com/digital-asset/daml/pull/3514/commits/e33087cc49371ab4073af90395ca9e2c59142654

**Pull request merged**